File: js/spec/polygon.spec.js
const GIS = require('../')

describe('gis.protobuf.Polygon2D has getGeoJSON function', function() {
  it('which returns a GeoJSON object', function() {
    let point1 = new GIS.Point2D(); point1.setX(1); point1.setY(2)
    let point2 = new GIS.Point2D(); point2.setX(3); point2.setY(4)

    let polygon = new GIS.Polygon2D(); polygon.setPointList([point1, point2])

    expect(polygon.toGeoJSON()).toEqual({
      type: 'Polygon',
      coordinates: [[[1, 2], [3, 4]]]
    })
  })
})

describe('gis.protobuf.Polygon3D has getGeoJSON function', function() {
  it('which returns a GeoJSON object', function() {
    let point1 = new GIS.Point3D(); point1.setX(1); point1.setY(2); point1.setZ(3)
    let point2 = new GIS.Point3D(); point2.setX(4); point2.setY(5); point2.setZ(6)

    let polygon = new GIS.Polygon3D(); polygon.setPointList([point1, point2])

    expect(polygon.toGeoJSON()).toEqual({
      type: 'Polygon',
      coordinates: [[[1, 2, 3], [4, 5, 6]]]
    })
  })
})

describe('gis.protobuf.MultiPolygon2D has getGeoJSON function', function() {
  it('which returns a GeoJSON object', function() {
    let point1 = new GIS.Point2D(); point1.setX(1); point1.setY(2)
    let point2 = new GIS.Point2D(); point2.setX(3); point2.setY(4)
    let polygon1 = new GIS.Polygon2D(); polygon1.setPointList([point1, point2])

    let point3 = new GIS.Point2D(); point3.setX(5); point3.setY(6)
    let point4 = new GIS.Point2D(); point4.setX(7); point4.setY(8)
    let polygon2 = new GIS.Polygon2D(); polygon2.setPointList([point3, point4])

    let multiPolygon = new GIS.MultiPolygon2D(); multiPolygon.setPolygonList([polygon1, polygon2])

    expect(multiPolygon.toGeoJSON()).toEqual({
      type: 'MultiPolygon',
      coordinates: [[[[1, 2], [3, 4]]], [[[5, 6], [7, 8]]]]
    })
  })
})

describe('gis.protobuf.MultiPolygon3D has getGeoJSON function', function() {
  it('which returns a GeoJSON object', function() {
    let point1 = new GIS.Point3D(); point1.setX(1); point1.setY(2); point1.setZ(3)
    let point2 = new GIS.Point3D(); point2.setX(4); point2.setY(5); point2.setZ(6)
    let polygon1 = new GIS.Polygon3D(); polygon1.setPointList([point1, point2])

    let point3 = new GIS.Point3D(); point3.setX(7); point3.setY(8); point3.setZ(9)
    let point4 = new GIS.Point3D(); point4.setX(10); point4.setY(11); point4.setZ(12)
    let polygon2 = new GIS.Polygon3D(); polygon2.setPointList([point3, point4])

    let multiPolygon = new GIS.MultiPolygon3D(); multiPolygon.setPolygonList([polygon1, polygon2])

    expect(multiPolygon.toGeoJSON()).toEqual({
      type: 'MultiPolygon',
      coordinates: [[[[1, 2, 3], [4, 5, 6]]], [[[7, 8, 9], [10, 11, 12]]]]
    })
  })
})

File: ruby/spec/gis/protobuf/polygon_spec.rb
require "spec_helper"

RSpec.describe "Gis::Protobuf::Polygon2D has getGeoJSON function" do
  it "which returns a GeoJSON object" do
    point = Gis::Protobuf::Polygon2D.new point: [
      Gis::Protobuf::Point2D.new(x: 1.0, y: 2.0),
      Gis::Protobuf::Point2D.new(x: 3.0, y: 4.0),
    ]

    expect(point.toGeoJSON()).to eq type: "Polygon", coordinates: [[[1.0, 2.0], [3.0, 4.0]]]
  end
end

RSpec.describe "Gis::Protobuf::Polygon3D has getGeoJSON function" do
  it "which returns a GeoJSON object" do
    point = Gis::Protobuf::Polygon3D.new point: [
      Gis::Protobuf::Point3D.new(x: 1.0, y: 2.0, z: 3.0),
      Gis::Protobuf::Point3D.new(x: 4.0, y: 5.0, z: 6.0),
    ]

    expect(point.toGeoJSON()).to eq type: "Polygon", coordinates: [[[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]]
  end
end

RSpec.describe "Gis::Protobuf::MultiPolygon2D has getGeoJSON function" do
  it "which returns a GeoJSON object" do
    point = Gis::Protobuf::MultiPolygon2D.new polygon: [
      Gis::Protobuf::Polygon2D.new(point: [
        Gis::Protobuf::Point2D.new(x: 1.0, y: 2.0),
        Gis::Protobuf::Point2D.new(x: 3.0, y: 4.0),
      ]),
      Gis::Protobuf::Polygon2D.new(point: [
        Gis::Protobuf::Point2D.new(x: 5.0, y: 6.0),
        Gis::Protobuf::Point2D.new(x: 7.0, y: 8.0),
      ])
    ]

    expect(point.toGeoJSON()).to eq type: "MultiPolygon", coordinates: [[[[1.0, 2.0], [3.0, 4.0]]], [[[5.0, 6.0], [7.0, 8.0]]]]
  end
end

RSpec.describe "Gis::Protobuf::MultiPolygon3D has getGeoJSON function" do
  it "which returns a GeoJSON object" do
    point = Gis::Protobuf::MultiPolygon3D.new polygon: [
      Gis::Protobuf::Polygon3D.new(point: [
        Gis::Protobuf::Point3D.new(x: 1.0, y: 2.0, z: 3.0),
        Gis::Protobuf::Point3D.new(x: 4.0, y: 5.0, z: 6.0),
      ]),
      Gis::Protobuf::Polygon3D.new(point: [
        Gis::Protobuf::Point3D.new(x: 7.0, y: 8.0, z: 9.0),
        Gis::Protobuf::Point3D.new(x: 10.0, y: 11.0, z: 12.0),
      ])
    ]

    expect(point.toGeoJSON()).to eq type: "MultiPolygon", coordinates: [[[[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]], [[[7.0, 8.0, 9.0], [10.0, 11.0, 12.0]]]]
  end
end
